### PR TITLE
Add circuit elements for pl-drawing

### PR DESCRIPTION
**Pull request by mfsilva22**
Adding four new pl drawing elements, to help creating circuit questions:
- pl-resistor
- pl-battery
- pl-capacitor
- pl-switch
- pl-inductor

image

Includes one example in XC 101, and documentation.

### CODE REVIEW

File: apps/prairielearn/elements/pl-drawing/elements.py [hunk 1]
```diff
@@ -1727,6 +1727,51 @@ def get_attributes():
         ]
 
 
+class Coil(BaseElement):
+    def generate(el, data):
+        stroke_color = pl.get_color_attrib(el, "stroke-color", "black")
+        x1 = pl.get_float_attrib(el, "x1", drawing_defaults["x1"])
+        y1 = pl.get_float_attrib(el, "y1", drawing_defaults["y1"])
+        if "x2" in el.attrib and "y2" in el.attrib:
```
Comment (eliotwrobson) [inline apps/prairielearn/elements/pl-drawing/elements.py:1735]: Is this the same style used in other places in this file? Not sure if `pl.has_attrib` is more appropriate.

**Review by eliotwrobson — COMMENTED** [review]
Python seems fine (besides the normal amount of pain associated with this element).

File: apps/prairielearn/elements/pl-drawing/elements.py (hunk 1 above)
Comment (mfsilva22) [reply apps/prairielearn/elements/pl-drawing/elements.py:1735]: It is the same style. I just copied the code from other similar elements (for example Spring, Line, etc)

File: exampleCourse/questions/demo/drawing/resistorCapacitorCircuit/question.html [hunk 2]
```diff
@@ -0,0 +1,32 @@
+<pl-question-panel>
+
+    <p>
+        <pl-drawing width="440" height="260"  grid-size="0" answers-name="answer">
+            <pl-drawing-initial>
+                <pl-battery        x1={{params.pA.0}} y1={{params.pA.1}} x2={{params.pB.0}} y2={{params.pB.1}} label="{{params.V}} V"></pl-battery>
```
Comment (nwalters512) [inline exampleCourse/questions/demo/drawing/resistorCapacitorCircuit/question.html:6]: Unless there's a reason for all the spaces after the element name, I'd remove them.

File: exampleCourse/questions/demo/drawing/resistorCapacitorCircuit/question.html [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+                <pl-resistor        x1={{params.pB.0}} y1={{params.pB.1}} x2={{params.pC.0}} y2={{params.pC.1}} label="{{params.R1}}\\Omega"></pl-resistor>
```
Comment (nwalters512) [inline exampleCourse/questions/demo/drawing/resistorCapacitorCircuit/question.html:7]: As a best practice, attribute values should be enclosed in double quotes. It's not strictly necessary in this case, but it is needed if there are spaces, quotes, or angle brackets in the string, so we usually include them so folks don't have to think too hard if they're pattern-matching against our example questions.

File: exampleCourse/questions/element/drawingGallery/question.html [hunk 4]
```diff
@@ -273,5 +283,54 @@
...
+        <pl-switch x1="20" y1="120" angle="20" width="120" switch-angle="45" label="S_1" draw-pin="false"></pl-switch>
+        <pl-switch x1="20" y1="100" angle="0" width="120" switch-angle="0"></pl-switch>
+    </pl-drawing-initial>
+</pl-drawing>
+</p>
+
+<p>
+    <pl-drawing width="200" height="200" grid-size="20">
+        <pl-drawing-initial>
+            <pl-inductor ></pl-inductor>
```
Comment (nwalters512) [inline exampleCourse/questions/element/drawingGallery/question.html:328]:
```suggestion
            <pl-inductor></pl-inductor>
```

File: exampleCourse/questions/element/drawingGallery/question.html [hunk 5]
```diff
@@ -273,5 +283,54 @@
...
+        <pl-battery x1="40" y1="120" angle="30" width="80" stroke-color="red" label="20 V"></pl-battery>
+    </pl-drawing-initial>
+</pl-drawing>
+</p>
+
+<p>
+<pl-drawing width="200" height="200" grid-size="20">
+    <pl-drawing-initial>
+        <pl-switch x1="20" y1="20" x2="140" y2="20" label="A"></pl-switch>
+        <pl-switch x1="180" y1="60" angle="90" width="120" label="B" stroke-color="blue"  ></pl-switch>
```
Comment (nwalters512) [inline exampleCourse/questions/element/drawingGallery/question.html:318]:
```suggestion
        <pl-switch x1="180" y1="60" angle="90" width="120" label="B" stroke-color="blue"></pl-switch>
```

File: exampleCourse/questions/element/drawingGallery/question.html [hunk 6]
```diff
@@ -273,5 +283,54 @@
...
+        <pl-resistor x1="40" y1="80" x2="180" y2="120" stroke-color="red" label="14\\Omega"></pl-resistor>
+        <pl-resistor x1="80" y1="160" height="30" stroke-color="purple" label="30\\Omega" offsety="25"></pl-resistor>
+    </pl-drawing-initial>
+</pl-drawing>
+</p>
+
+<p>
+<pl-drawing width="200" height="200" grid-size="20">
+    <pl-drawing-initial>
+        <pl-battery x1="20" y1="40" x2="140" y2="40" label="10 V" ></pl-battery>
```
Comment (nwalters512) [inline exampleCourse/questions/element/drawingGallery/question.html:308]:
```suggestion
        <pl-battery x1="20" y1="40" x2="140" y2="40" label="10 V"></pl-battery>
```

File: exampleCourse/questions/element/drawingGallery/question.html [hunk 7]
```diff
@@ -273,5 +283,54 @@
...
+        <pl-capacitor x1="20" y1="40" x2="140" y2="40" label="20\\mu C" ></pl-capacitor>
+        <pl-capacitor x1="40" y1="180" x2="180" y2="100" stroke-color="blue" label="4\\mu C"></pl-capacitor>
+    </pl-drawing-initial>
+</pl-drawing>
+</p>
+
+<p>
+<pl-drawing width="200" height="200" grid-size="20">
+    <pl-drawing-initial>
+        <pl-resistor x1="20" y1="40" x2="140" y2="40" label="2\\Omega" ></pl-resistor>
```
Comment (nwalters512) [inline exampleCourse/questions/element/drawingGallery/question.html:298]:
```suggestion
        <pl-resistor x1="20" y1="40" x2="140" y2="40" label="2\\Omega"></pl-resistor>
```

File: exampleCourse/questions/element/drawingGallery/question.html [hunk 8]
```diff
@@ -273,5 +283,54 @@
 </pl-drawing>
 </p>
 
+<p>
+<pl-drawing width="200" height="200" grid-size="20">
+    <pl-drawing-initial>
+        <pl-capacitor x1="20" y1="40" x2="140" y2="40" label="20\\mu C" ></pl-capacitor>
```
Comment (nwalters512) [inline exampleCourse/questions/element/drawingGallery/question.html:289]:
```suggestion
        <pl-capacitor x1="20" y1="40" x2="140" y2="40" label="20\\mu C"></pl-capacitor>
```

File: exampleCourse/questions/element/drawingGallery/question.html [hunk 9]
```diff
@@ -131,6 +131,16 @@
   </pl-drawing>
 </p>
 
+<p>
+    <pl-drawing width="200" height="200" grid-size="20">
+        <pl-drawing-initial>
+            <pl-coil x1=50 y1=40  draw-pin="true"></pl-coil>
+            <pl-coil x1=20 y1=100 angle=20 width=120 height=80></pl-coil>
+            <pl-coil x1=20 y1=180 width=40 height=20 stroke-color="purple3"></pl-coil>
```
Comment (nwalters512) [inline exampleCourse/questions/element/drawingGallery/question.html:139]:
```suggestion
            <pl-coil x1="50" y1="40" draw-pin="true"></pl-coil>
            <pl-coil x1="20" y1="100" angle="20" width="120" height="80"></pl-coil>
            <pl-coil x1="20" y1="180" width="40" height="20" stroke-color="purple3"></pl-coil>
```

File: exampleCourse/questions/demo/drawing/resistorCapacitorCircuit/question.html [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+    </p>
+    
+</pl-question-panel>
+
+<p>
+    <pl-number-input answers-name="charge" suffix="$\mu C$"></pl-number-input>
+</p>
+
+
+
```
Comment (nwalters512) [inline exampleCourse/questions/demo/drawing/resistorCapacitorCircuit/question.html:32]: Too many trailing newlines.

```suggestion
```

File: apps/prairielearn/elements/pl-drawing/mechanicsObjects.js [hunk 11]
```diff
@@ -3596,6 +3695,461 @@ mechanicsObjects.byType['pl-controlled-curved-line'] = class extends PLDrawingBa
...
+        textAlign: 'left',
+        fontSize: options.fontSize,
+        selectable: false,
+        originX: 'center',
+        originY: 'center',
+      });
+      canvas.add(textObj);
+    }
+
+    return;
```
Comment (nwalters512) [inline apps/prairielearn/elements/pl-drawing/mechanicsObjects.js:3890]: There's no need for these empty `return` statements at the end of a function; they're implicit. Ditto for the other new element classes.

File: apps/prairielearn/elements/pl-drawing/mechanicsObjects.js [hunk 12]
```diff
@@ -3596,6 +3695,461 @@ mechanicsObjects.byType['pl-controlled-curved-line'] = class extends PLDrawingBa
...
+    return 'Add battery';
+  }
+};
+
+mechanicsObjects.byType['pl-resistor'] = class extends PLDrawingBaseElement {
+  static generate(canvas, options) {
+    var theta = Math.atan2(options.y2 - options.y1, options.x2 - options.x1);
+    var d = Math.sqrt(Math.pow(options.y2 - options.y1, 2) + Math.pow(options.x2 - options.x1, 2));
+
+    function getPositions(gap, d, theta, options) {
```
Comment (nwalters512) [inline apps/prairielearn/elements/pl-drawing/mechanicsObjects.js:3903]: What does this function do? More broadly, I'd love to see explanations for what all the code for these new elements is doing. At the very least, `obj1`, `obj2`, `obj3`, etc. should have meaningful names that give me at least some sense of what's happening here.

File: exampleCourse/questions/demo/drawing/resistorCapacitorCircuit/question.html (hunk 3 above)
Comment (mfsilva22) [reply exampleCourse/questions/demo/drawing/resistorCapacitorCircuit/question.html]: got it!

**Review by nwalters512 — APPROVED** [review]

**Pull request merged**